### PR TITLE
 #167765394 User can login via google  and linkedin

### PR DESCRIPTION
**Pull request by voke1**
### What does this PR do?

- User can login via google  and linkedin

#### Sign in via Linkedin:
- Register app with  Linkedin to get config credentials
- Write route for Login via Linkedin
- Write a controller function to return Linkedin connection url.
- Write a callback/redirect route for Linkedin.
- Create a controller function to handle get request from Linkedin API, upon user successful login.
- Obtain logged in user email.
- Check if user email is present in WeTravel database, return success code and user email. If absent, 
    return error code and error message.
- Write unit tests
- Document using swagger

#### Sign in via Google:
- Register app with Google to get config credentials
- Write route for Login via Google
- Write a controller function to return Google connection url.
- Write a callback/redirect route for Google redirect.
- Create a controller function to handle get request from Google API, upon user successful login.
- Obtain logged in user email.
- Check if user email is present in WeTravel database, return success code and user email. If absent, 
  return error code and error message.
- Write unit tests
- Document using swagger

### How should this be manually tested?
- Run  `npm run test` for automated test

#### To test Google login	
- Run `npm run reset:db`
- Run `npm run start`
- Send a GET request to /api/v1/auth/user/google/signin
- Enter url returned in browsers to log in with google

#### To test Linkedin login
- Run `npm reset:db`
- Run `npm run start`
- Send a GET request to /api/v1/auth/user/linkendin/signin
- Enter url returned in browsers to log in with Linkedin

### Any background context you want to provide?

 Please use the below test email for a successful login to WeTravel for both Google and Linkedin:
   - Account: wetravel.user1@gmail.com
   - Password: wetravelpassword!

### What are the relevant pivotal tracker stories?

- #167765394

### Implementation Plan
- https://drive.google.com/file/d/16zHSmHKyR4YC4kuW91DnBrlEt2kV-TFg/view?usp=sharing

### CODE REVIEW

File: src/controllers/userController.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
+import https from 'https';
+import userHelper from '../helpers/userHelper';
+
+const scope = ['r_liteprofile', 'w_member_social', 'r_emailaddress'];
+const clientId = process.env.LINKENDIN_CLIENT_ID;
+const clientSecret = process.env.LINKENDIN_CLIENT_SECRET;
+const redirect = process.env.LINKENDIN_REDIRECT_URL;
+const Linkedin = require('node-linkedin')(clientId, clientSecret, redirect);
```
Comment (debelistic) [inline src/controllers/userController.js]: Why are using require?

File: src/helpers/userHelper.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import { google } from 'googleapis';
+import dotenv from 'dotenv';
```
Comment (debelistic) [inline src/helpers/userHelper.js]: Instead of importing the entire module why not `import { config } from dotenv`
Then call it as `config`

Comment (voke1) [reply src/helpers/userHelper.js]: Noted

File: src/controllers/userController.js (hunk 1 above)
Comment (voke1) [reply src/controllers/userController.js]: Noted

File: src/controllers/userController.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      * @return {json} Returns json object
+      * @memberof User
+      */ 
+  async linkedinUrl(req, res) {
+    this.authurl = Linkedin.auth.authorize(scope);
+    const url = this.authurl;
+    return res.status(200).send({ status: 200, url });
+  }
+
+  // Take the "code" parameter which Linkedin gives us once when the user logs in, get an access token,
```
Comment (misanj) [inline src/controllers/userController.js]: You can do  `@description` in your `js docs` to describe what a particular method or function does... Instead of commenting on the file.

File: src/database/config.js [hunk 4]
```diff
@@ -4,29 +4,29 @@ module.exports = {
...
+    dialect: "postgres",
     logging: false,
   },
   test: {
     username: process.env.DB_USER,
     password: process.env.DB_PASSWORD,
     database: process.env.DB_NAME,
     host: process.env.DB_HOST,
-    dialect: 'postgres',
+    dialect: "postgres",
```
Comment (misanj) [inline src/database/config.js]: why are you using double quotes here?

File: .envSample [hunk 5]
```diff
@@ -7,3 +7,9 @@ DB_USER=
 DB_PASSWORD=
 DB_NAME=
 DB_HOST=
+GOOGLE_CLIENT_ID = 
+GOOGLE_CLIENT_SECRET = 
+GOOGLE_REDIRECT_URL = 
+LINKENDIN_CLIENT_ID = 
+LINKENDIN_CLIENT_SECRET = 
+LINKENDIN_REDIRECT_URL=
```
Comment (misanj) [inline .envSample]: add a new line here

File: src/controllers/userController.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
+/* eslint-disable linebreak-style */
+import https from 'https';
+import userHelper from '../helpers/userHelper';
+import models from '../models';
+
+const scope = ['r_liteprofile', 'w_member_social', 'r_emailaddress'];
+const clientId = process.env.LINKENDIN_CLIENT_ID;
+const clientSecret = process.env.LINKENDIN_CLIENT_SECRET;
+const redirect = process.env.LINKENDIN_REDIRECT_URL;
+const Linkedin = require('node-linkedin')(clientId, clientSecret, redirect);
```
Comment (misanj) [inline src/controllers/userController.js]: You're still using require here! Is there a reason for that?

File: src/controllers/userController.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+      * @return {json} Returns json object
+      * @memberof User
+      */ 
+  async getLinkedinAccountFromCode(req, res) {
+    Linkedin.auth.getAccessToken(
+      res,
+      req.query.code,
+      req.query.state,
+      async (err, results) => {
+        if (err) return console.error(err);
```
Comment (misanj) [inline src/controllers/userController.js]: remove this console here and at line `112`

File: src/controllers/userController.js [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+import https from 'https';
+import userHelper from '../helpers/userHelper';
+import models from '../models';
+
+const scope = ['r_liteprofile', 'w_member_social', 'r_emailaddress'];
+const clientId = process.env.LINKENDIN_CLIENT_ID;
+const clientSecret = process.env.LINKENDIN_CLIENT_SECRET;
+const redirect = process.env.LINKENDIN_REDIRECT_URL;
+const Linkedin = require('node-linkedin')(clientId, clientSecret, redirect);
+
```
Comment (misanj) [inline src/controllers/userController.js]: Eslint has flag so much error on this file....I had to resolve most of them but could not test your endpoint yet

**Review by misanj — COMMENTED** [review]
Good job @voke1 please implement the changes on the highlighted lines and fix eslint errors, I don't know if am the only one seeing them. I get this error when I tried to start the server `Error: Cannot find module 'googleapis'` maybe I need to set environmental variables in `.env ` file first

File: src/controllers/userController.js (hunk 3 above)
Comment (voke1) [reply src/controllers/userController.js]: Thanks. I'll remove inline comments

File: src/controllers/userController.js (hunk 7 above)
Comment (voke1) [reply src/controllers/userController.js]: Noted

**Comment by voke1** [comment]
> Good job @voke1 please implement the changes on the highlighted lines and fix eslint errors, I don't know if am the only one seeing them. I get this error when I tried to start the server `Error: Cannot find module 'googleapis'` maybe I need to set environmental variables in `.env ` file first

Thanks for your feedback. You will need to however run `npm install` to resolve the issue with `Cannot find module googleapis`. Kindly go-ahead to further test functionality.

Also, I'll need to update this branch to align with recent changes made on the develop branch and resolve conflicting files.

File: src/controllers/userController.js (hunk 8 above)
Comment (voke1) [reply src/controllers/userController.js]: Noted. I'll check on this

File: .envSample (hunk 5 above)
Comment (voke1) [reply .envSample]: Thanks. I'll implement this

File: src/database/config.js (hunk 4 above)
Comment (voke1) [reply src/database/config.js]: Noted

File: src/controllers/userController.js (hunk 8 above)
Comment (victor-nach) [reply src/controllers/userController.js]: yes, there a couple errors on eslint

File: src/controllers/userController.js [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+  /**
+      * @param {*} req Request object
+      * @param {*} res Response object
+      * @return {json} Returns json object
+      * @memberof User
+      */ async googleUrl(req, res) {
+    try {
+      this.auth = userHelper.createConnection();
+      const url = userHelper.getConnectionUrl(this.auth);
+      return res.status(200).send({ status: 200, url });
```
Comment (victor-nach) [inline src/controllers/userController.js]: why are you using res.send() instead of res.json()

File: src/controllers/userController.js [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+            return res.status(200).json({ status: 200, user });
+          });
+        });
+        profileRequest.end();
+      }
+    );
+  }
+}
+
+export default new User();
```
Comment (victor-nach) [inline src/controllers/userController.js]: we already agreed to use static methods and export classes, I think it's better if you try to maintain consistency, however check that it doesn't affect your usage of 'this'

File: src/models/user.js [hunk 11]
```diff
@@ -1,28 +1,41 @@
...
-    email: {
-      allowNull: false,
-      type: DataTypes.STRING,
-      unique: true
-    },
-    phoneNumber: {
-      allowNull: false,
-      type: DataTypes.STRING
+  const User = sequelize.define(
+    "User",
```
Comment (victor-nach) [inline src/models/user.js]: double quotes, do you have eslint installed on your vscode?

**Comment by anneKay** [comment]
Please, can you attend to reviews on this PR and resolve your conflicts. 

Thanks. This branch has been updated and all reviews implemented.

File: src/controllers/UserController.js [hunk 12]
```diff
@@ -27,4 +36,102 @@ export default class UserController {
...
+   * @param {object} req The request object
+   * @param {object} res The response object
+   * @returns {object} The API response
+   * @memberof User
+   */
+  static async getGoogleUrl(req, res) {
+    try {
+      const auth = Utils.createConnection();
+      const url = Utils.getConnectionUrl(auth);
+      return res.status(200).json({ status: 200, url });
```
Comment (okobaba1) [inline src/controllers/UserController.js]: Make use of the response middleware
this should look like:  `return resSuccess(res, 200, url);`

File: src/controllers/UserController.js [hunk 13]
```diff
@@ -27,4 +36,102 @@ export default class UserController {
...
+    } catch (error) {
+      return res.status(500).json({ status: 500, error: error.message });
```
Comment (okobaba1) [inline src/controllers/UserController.js]: and this like: `return resError(res, 500, error.message);`

File: src/controllers/UserController.js [hunk 14]
```diff
@@ -27,4 +36,102 @@ export default class UserController {
...
+      const auth = Utils.createConnection();
+      const data = await auth.getToken(req.query.code);
+      const token = data.tokens;
+      auth.setCredentials(token);
+      const plus = Utils.getGooglePlusApi(auth);
+      const me = await plus.people.get({ userId: 'me' });
+      const email = me.data.emails && me.data.emails.length && me.data.emails[0].value;
+      const user = await UserServices.getUserByGmail(email);
+      if (!user) {
+        return res.status(404).json({ status: 404, message: 'User not found' });
```
Comment (okobaba1) [inline src/controllers/UserController.js]: here also

File: src/controllers/userController.js (hunk 6 above)
Comment (voke1) [reply src/controllers/userController.js]: This has been updated

File: src/controllers/userController.js (hunk 9 above)
Comment (voke1) [reply src/controllers/userController.js]: This has been updated

File: public/api-docs/swaggerDoc.json [hunk 15]
```diff
@@ -177,9 +177,129 @@
...
+                ],
+                "responses": {
+                    "200": {
+                        "description": "successful operation",
+                        "schema": {
+                            "$ref": "#/definitions/User"
+                        }
                     },
-                    "422": {
-                        "description": "Invalid Input"
```
Comment (anneKay) [inline public/api-docs/swaggerDoc.json:182]: Why did you remove this?

File: src/models/user.js (hunk 11 above)
Comment (voke1) [reply src/models/user.js]: I actually do. Thanks. This has been resolved.

File: src/controllers/userController.js (hunk 10 above)
Comment (voke1) [reply src/controllers/userController.js]: Noted

File: public/api-docs/swaggerDoc.json (hunk 15 above)
Comment (voke1) [reply public/api-docs/swaggerDoc.json:182]: I actually replaced it with ```  "404": {
                        "description": "User not found"```. Since it doesn't describe the error response of the specified endpoint.

Comment (okobaba1) [reply public/api-docs/swaggerDoc.json:182]: It does describe errors from the input, mostly errors from express-validator

**Comment by voke1** [comment]
@okobaba1 Thanks for pointing that out. However, the `422` error code wasn't specified in any of my enpoints, instead, I used `404` as appropriate. Since I'm only using a GET request I believe there are no input parameters to validate.

File: src/controllers/UserController.js (hunk 12 above)
Comment (voke1) [reply src/controllers/UserController.js]: Noted

File: src/controllers/UserController.js (hunk 13 above)
Comment (voke1) [reply src/controllers/UserController.js]: Okay

File: public/api-docs/swaggerDoc.json (hunk 15 above)
Comment (voke1) [reply public/api-docs/swaggerDoc.json:182]: @okobaba1 Thanks for pointing that out. However, the 422 error code wasn't specified in any of my enpoints, instead, I used 404 as appropriate. Since I'm only using a GET request I believe there are no input parameters to validate.

Comment (okobaba1) [reply public/api-docs/swaggerDoc.json:182]: I think you removed it from the signup.

Comment (voke1) [reply public/api-docs/swaggerDoc.json:182]: No. This is what is currently present in signup ``` "409": {
                        "description": "User already exists"
                    }```

Comment (anneKay) [reply public/api-docs/swaggerDoc.json:182]: 409 and 422 error codes were both present in signup, please add the 422 description you removed
